src/platform/Zephyr/KeyValueStoreManagerImpl.cpp

```diff
@@ -152,26 +108,126 @@ void KeyValueStoreManagerImpl::Init()
     VerifyOrDie(settings_subsys_init() == 0);
 }
 
+bool KeyValueExist(const char * fullkey, size_t * val_len)
+{
+    ssize_t len;
+    uint8_t emptyValue[kEmptyValueSize];
+
+    len = settings_get_val_len(fullkey);
+    VerifyOrReturnError(len > 0, false);
+
+    if (len == kEmptyValueSize)
+    {
+        len = settings_load_one(fullkey, emptyValue, kEmptyValueSize);
+        if ((len == kEmptyValueSize) && !memcmp(emptyValue, kEmptyValue, kEmptyValueSize))
+        {
+            *val_len = 0;
+        }
+        else
+        {
+            *val_len = len;
+        }
+    }
+
+    return true;
+}
+
+void LoadOneAndVerifyResult(const char * fullkey, void * dest_buf, size_t dest_size, size_t * readSize,
+                               CHIP_ERROR * result)
+{
+    ssize_t bytesRead = settings_load_one(fullkey, dest_buf, dest_size);
+
+    // If the return code is -ENOENT the key is not found
+    if ((bytesRead == -ENOENT) || (!bytesRead))
+    {
+        *result = CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+        *readSize = 0;
+        return;
+    }
+    if ((bytesRead == kEmptyValueSize) && !memcmp(dest_buf, kEmptyValue, kEmptyValueSize))
+    {
+        *result = CHIP_NO_ERROR;
+        *readSize = 0;
+        return;
+    }
+    if ((size_t)bytesRead > dest_size)
+    {
+        *result = CHIP_ERROR_BUFFER_TOO_SMALL;
+        *readSize = dest_size;
+        return;
+    }
+    else if (bytesRead >= 0)
+    {
+        *result = CHIP_NO_ERROR;
+    }
+    else
+    {
+        *result = CHIP_ERROR_PERSISTED_STORAGE_FAILED;
+    }
+
+    *readSize = (bytesRead > 0) ? bytesRead : 0;
+
+    return;
+}
+
 CHIP_ERROR KeyValueStoreManagerImpl::_Get(const char * key, void * value, size_t value_size, size_t * read_bytes_size,
                                           size_t offset_bytes) const
 {
+    CHIP_ERROR result;
+    size_t readSize = 0;
+    uint8_t emptyValue[kEmptyValueSize];
     // Offset and partial reads are not supported, for now just return NOT_IMPLEMENTED.
     // Support can be added in the future if this is needed.
     VerifyOrReturnError(offset_bytes == 0, CHIP_ERROR_NOT_IMPLEMENTED);
 
     char fullKey[SETTINGS_MAX_NAME_LEN + 1];
     ReturnErrorOnFailure(MakeFullKey(fullKey, key));
 
-    ReadEntry entry{ value, value_size, 0, CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND };
-    settings_load_subtree_direct(fullKey, LoadEntryCallback, &entry);
+    if ((!value) || (value_size == 0))
+    {
+        // we want only to verify that the key exists
+        if (!KeyValueExist(fullKey, &readSize))
+        {
+            result = CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+            readSize = 0;
+        }
+        else
+        {
+            result = CHIP_NO_ERROR;
```

Comment:
This is wrong. We need to return `CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND` when an empty value is saved (which my original patch did).

Reply:
ah, nevermind, this is OK.